### PR TITLE
Rc/track bulk group updates

### PR DESCRIPTION
**Pull request by Robert-Costello**
## Technical Summary

Tracks changes to groups (via bulk upload) to `UserHistory`
[USH-1314](https://dimagi-dev.atlassian.net/browse/USH-1314)

## Feature Flag

`USER_HISTORY_REPORT`
### Safety story

Tested locally, changes are behind feature flag not currently in use on prod
### QA Plan
not requesting QA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/user_importer/importer.py [hunk 1]
```diff
@@ -582,35 +580,60 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
+                    updated_groups = []
+                    for group in user_curr_groups:
+                        if group not in removed_from_groups:
+                            updated_groups.append(group)
+                    updated_groups.extend(user_added_to_groups)
+                    commcare_user_importer.logger.add_info(
+                        UserChangeMessage.groups_info(updated_groups)
+                    )
+
+                commcare_user_importer.save_log()
```
Comment (snopoke) [inline corehq/apps/user_importer/importer.py:627]: This seems quite complex. I think it could be simplified:
```python
old_group_ids = set()
for group in domain_info.group_memoizer.by_user_id(user.user_id):
    old_group_ids.add(group.get_id)
    if group.name not in group_names:
        group.remove_user(user)

new_groups = {}
for group_name in group_names:
    group = domain_info.group_memoizer.by_name(group_name)
    group.add_user(user, save=False)
    new_groups[group.get_id] = group

if set(new_groups) != old_group_ids:
    commcare_user_importer.logger.add_info(UserChangeMessage.groups_info(list(new_groups.values()))
```

I think it would also be good to put this in a function on the user importer class:
```
error = commcare_user_importer.update_user_groups(domain_info, group_names)
if error:
    ret['errors'].append(error)

File: corehq/apps/user_importer/importer.py [hunk 2]
```diff
@@ -582,35 +578,63 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
-                "and try again."
-            )
-            logging.exception((
-                'BulkSaveError saving groups. '
-                'User saw error message "%s". Errors: %s'
-            ) % (_error_message, e.errors))
-            ret['errors'].append(_error_message)
-
-    return ret
+        return ret
```
Comment (mkangia) [inline corehq/apps/user_importer/importer.py:636]: is this correct indentation?

**Comment by mkangia** [comment]
@Robert-Costello Would be good to add tests. User import is pretty delicate considering the size of that method that does the import.

File: corehq/apps/user_importer/importer.py (hunk 1 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/importer.py:627]: You're right, it is pretty complex, thanks for the suggestion on simplifying. Moving the logic to `CommCareUserImporter` also makes a lot of sense.

File: corehq/apps/user_importer/importer.py (hunk 2 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/importer.py:636]: I think it's correct. It caught my eye too when I first moved it into that `finally` block, but all local and automated testing checks out.

**Comment by Robert-Costello** [comment]
@mkangia Agreed, I'll add some tests.

File: corehq/apps/user_importer/importer.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/user_importer/importer.py:610]: since there is no code in the `finally` block now you could remove the entire `try / finally`. Having the return statement in there will not make the function return the data if an exception is thrown.

File: corehq/apps/user_importer/helpers.py [hunk 3]
```diff
@@ -252,6 +252,25 @@ def update_locations(self, location_codes, domain_info):
...
+            new_groups = {}
+            for group_name in group_names:
+                group = domain_info.group_memoizer.by_name(group_name)
+                group.add_user(self.user, save=False)
+                new_groups[group.get_id] = group
+
+            if set(new_groups) != old_group_ids:
+                self.logger.add_info(UserChangeMessage.groups_info(list(new_groups.values())))
+        except Exception as e:
+            return str(e)
```
Comment (snopoke) [inline corehq/apps/user_importer/helpers.py:272]: what exceptions can get raised here? I had imagined that the error would come from the `BulkSaveError` when calling `domain_info.group_memoizer.save_all()`.

Thinking about this a bit more I can see why that was in the finally block - so that we don't re-save the groups for after each user row is processed.

File: corehq/apps/user_importer/importer.py [hunk 4]
```diff
@@ -582,35 +579,35 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
 
-                for group in domain_info.group_memoizer.by_user_id(user.user_id):
-                    if group.name not in group_names:
-                        group.remove_user(user)
+                error = commcare_user_importer.update_user_groups(domain_info, group_names)
+                if error:
+                    error_message = f"{user.user_id}: {error}"
+                    ret['errors'].append(_(error_message))
+                try:
+                    for domain_info in domain_info_by_domain.values():
```
Comment (snopoke) [inline corehq/apps/user_importer/importer.py:587]: I think this should probably stay in the `finally` block to prevent re-saving the groups for every user row processed. I guess the tricky thing now is that the log is saved in the loop but the groups are saved out of the loop.

One option would be to make the memoizer smarter so that it knows which groups have been updated and need to be saved. That way we could do the save in the loop and avoid having to save all the groups every time. It would still be more expensive than doing the save at the end but we'd have better consistency between the data and the logs.

File: corehq/apps/user_importer/importer.py [hunk 5]
```diff
@@ -582,35 +579,35 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
                     # Passing use_primary_db=True because of https://dimagi-dev.atlassian.net/browse/ICDS-465
                     user.get_django_user(use_primary_db=True).check_password(password)
 
-                for group in domain_info.group_memoizer.by_user_id(user.user_id):
-                    if group.name not in group_names:
-                        group.remove_user(user)
+                error = commcare_user_importer.update_user_groups(domain_info, group_names)
+                if error:
+                    error_message = f"{user.user_id}: {error}"
```
Comment (mkangia) [inline corehq/apps/user_importer/importer.py:584]: I don't think you need to add user_id to the message, the error message will already be bound to the user row in `ret` no?
Plus, if we need to add a user identity here, we should use username so that when the users see it, they can link it to a user.

File: corehq/apps/user_importer/importer.py [hunk 6]
```diff
@@ -582,35 +579,35 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
+                        "Please make sure no one else is editing your groups "
+                        "and try again."
+                    )
+                    logging.exception((
+                        'BulkSaveError saving groups. '
+                        'User saw error message "%s". Errors: %s'
+                    ) % (_error_message, e.errors))
+                    ret['errors'].append(_error_message)
+
+                commcare_user_importer.save_log()
```
Comment (mkangia) [inline corehq/apps/user_importer/importer.py:601]: @Robert-Costello I am a bit concerned by the order of actions here. `commcare_user_importer.save_log()` is called right in the end, way after `user.save`.
My understanding is that there is a possibility that user gets updated but the log is not saved?
Suggesting if we could keep the log save where it is and just update the change_messages if group update is successful?
Also, even currently, we have logged group updates but they can fail at the bulk update step, leading to an incorrect logging.

File: corehq/apps/user_importer/helpers.py [hunk 7]
```diff
@@ -252,6 +252,25 @@ def update_locations(self, location_codes, domain_info):
...
+
+            new_groups = {}
+            for group_name in group_names:
+                group = domain_info.group_memoizer.by_name(group_name)
+                group.add_user(self.user, save=False)
+                new_groups[group.get_id] = group
+
+            if set(new_groups) != old_group_ids:
+                self.logger.add_info(UserChangeMessage.groups_info(list(new_groups.values())))
+        except Exception as e:
```
Comment (mkangia) [inline corehq/apps/user_importer/helpers.py:271]: Can you share why we have a generic exception catch here?

File: corehq/apps/user_importer/importer.py (hunk 5 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/importer.py:584]: Yeah, that makes sense. I'll take out the user_id.

File: corehq/apps/user_importer/helpers.py (hunk 7 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/helpers.py:271]: The generic exception is my Node/JavaScript instinct. It's more Pythonic to anticipate specific exceptions and I'll get into the habit of doing it that way. Perhaps we don't need to handle a possible error here since `update_user_groups` is called inside the try block in `create_or_update_commcare_users_and_groups`. What do you think?

Comment (mkangia) [reply corehq/apps/user_importer/helpers.py:271]: Not sure about the comparison between languages, but generally I think Generic exceptions are needed in very rare scenarios. 
And ya, good point about the parent try block. In this block of code if we don't expect an exception, we should not have an except (`except when expect` if you would say :) )

File: corehq/apps/user_importer/importer.py (hunk 6 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/importer.py:601]: Ok, I've put `commcare_user_importer.save_log()` back with `user.save()`. `commcare_user_importer.save_log` now returns the record. I modified `update_user_groups` to make the group updates and return new change messages. Once bulk upload has succeeded I'm updating the record with the new change messages. Haven't had a chance to add automated tests yet, but it does appear to work when testing locally.

File: corehq/apps/user_importer/importer.py (hunk 4 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/importer.py:587]: I added a set to track group changes to the memoizer. It's populated for each user in the loop by `update_user_groups`,  saved and then cleared for the next user.

File: corehq/apps/user_importer/importer.py [hunk 8]
```diff
@@ -437,178 +437,182 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
+
+            group_change_message = {}
+            if group_names:
+                group_change_message = commcare_user_importer.update_user_groups(domain_info, group_names)
+
+            try:
+                for group_id in domain_info.group_memoizer.updated_groups:
+                    group = domain_info.group_memoizer.groups_by_id[group_id]
+                    group.save()
+                domain_info.group_memoizer.updated_groups.clear()
```
Comment (snopoke) [inline corehq/apps/user_importer/importer.py:592]: it would be great if this was encapsulated in the memoizer:
```python
def save_updated(self):
    updated = [self.groups_by_id[_id] for _id in self.updated_groups]
    Group.bulk_save(updated)
    self.updated_groups.clear()
```

File: corehq/apps/user_importer/helpers.py [hunk 9]
```diff
@@ -252,6 +252,26 @@ def update_locations(self, location_codes, domain_info):
...
+        for group in domain_info.group_memoizer.by_user_id(self.user.user_id):
+            old_group_ids.add(group.get_id)
+            if group.name not in group_names:
+                group.remove_user(self.user)
+                domain_info.group_memoizer.updated_groups.add(group.get_id)
+
+        new_groups = {}
+        for group_name in group_names:
+            group = domain_info.group_memoizer.by_name(group_name)
+            group.add_user(self.user, save=False)
```
Comment (snopoke) [inline corehq/apps/user_importer/helpers.py:266]: can we make `group.add_user` return `True` if the group changed? Then this could become:
```python
if group.add_user(self.user, save=False):
    domain_info.group_memoizer.group_updated(group)
```

**Review by snopoke — COMMENTED** [review]
nearly there, a few more comments that I think will clean things up a bit

File: corehq/apps/user_importer/importer.py (hunk 8 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/importer.py:592]: That makes sense. Just got around to finishing making this change. When I was working on this yesterday I realized that I needed to account for changes to groups alone. Previously group changes were dependent on a `log` being returned that could then be updated. That wasn't happening unless another change to the user was made first.

File: corehq/apps/user_importer/helpers.py (hunk 9 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/helpers.py:280]: The idea is to have `group.add_user` tell us if the group changed instead of checking against `old_group_ids`? That makes sense to me. What about `group_memoizer.group_updated(group)`? It's not currently a method on the memoizer. Did you have something in mind for it, or does `group_memoizer.updated_groups.add(group.get_id)` still work?

File: corehq/apps/user_importer/importer.py [hunk 10]
```diff
@@ -587,8 +587,7 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
                 user.get_django_user(use_primary_db=True).check_password(password)
 
             group_change_message = {}
```
Comment (snopoke) [inline corehq/apps/user_importer/importer.py:591]: no need to have this here anymore

File: corehq/apps/user_importer/importer.py [hunk 11]
```diff
@@ -532,6 +532,8 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
             if role:
                 role_qualified_id = domain_info.roles_by_name[role]
                 commcare_user_importer.update_role(role_qualified_id)
+            else:
+                commcare_user_importer.update_role('none', clear_role=True)
```
Comment (snopoke) [inline corehq/apps/user_importer/importer.py:536]: I know you didn't make it this way but it's weird to pass in `'none'` as a string

File: corehq/apps/user_importer/helpers.py [hunk 12]
```diff
@@ -142,10 +142,12 @@ def __init__(self, upload_domain, user_domain, user, upload_user, is_new_user, v
 
         self.role_updated = False
 
-    def update_role(self, role_qualified_id):
+    def update_role(self, role_qualified_id, clear_role=False):
         user_current_role = self.user.get_role(domain=self.user_domain)
         self.role_updated = not (user_current_role
                                  and user_current_role.get_qualified_id() == role_qualified_id)
+        if user_current_role and clear_role:
```
Comment (snopoke) [inline corehq/apps/user_importer/helpers.py:149]: I don't think you need to pass in `clear_role`.  When you pass in `'none'` as the `role_qualified_id` then `self.role_updated` is going to get set to `True` in the line above.

File: corehq/apps/user_importer/helpers.py [hunk 13]
```diff
@@ -252,6 +268,26 @@ def update_locations(self, location_codes, domain_info):
...
+            group = domain_info.group_memoizer.by_name(group_name)
+            new_groups[group.get_id] = group
+            if group.add_user(self.user, save=False):
+                domain_info.group_memoizer.updated_groups.add(group.get_id)
+
+        if set(new_groups) != old_group_ids:
+            return UserChangeMessage.groups_info(list(new_groups.values()))
+
+
+
```
Comment (snopoke) [inline corehq/apps/user_importer/helpers.py:290]: extra blank lines

File: corehq/apps/user_importer/importer.py [hunk 14]
```diff
@@ -437,178 +442,182 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
+                # Other than that, I'm not sure what's going on
+                # Passing use_primary_db=True because of https://dimagi-dev.atlassian.net/browse/ICDS-465
+                user.get_django_user(use_primary_db=True).check_password(password)
+
+            group_change_message = {}
+            group_change_message = commcare_user_importer.update_user_groups(domain_info, group_names)
+
+            try:
+                domain_info.group_memoizer.save_updated()
+
```
Comment (snopoke) [inline corehq/apps/user_importer/importer.py:596]: blank line

**Review by snopoke — COMMENTED** [review]
Almost there!

File: corehq/apps/user_importer/helpers.py (hunk 9 above)
Comment (snopoke) [reply corehq/apps/user_importer/helpers.py]: it would still work to do that but you're unnecessarily exposing the `updated_groups` property of the memoizer. Making it a method [encapsulates](https://gamedevelopment.tutsplus.com/tutorials/quick-tip-the-oop-principle-of-encapsulation--gamedev-2187) the functionality in the class.

File: corehq/apps/user_importer/importer.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/user_importer/importer.py]: I was wrong about the `return` statement FYI. If you `return` from inside a finally block when an exception has been raised then the exception will get silenced which is even worse.

File: corehq/apps/user_importer/helpers.py (hunk 12 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/helpers.py:149]: You're right, removing `clear_role`.

File: corehq/apps/user_importer/importer.py (hunk 11 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/importer.py:536]: It did seem odd. Does it make sense to change it is this PR, to `None`? I think `_AuthoizableMixin` would need to be updated and possibly some methods in other places like. Should that be a followup PR?

File: corehq/apps/user_importer/helpers.py (hunk 9 above)
Comment (Robert-Costello) [reply corehq/apps/user_importer/helpers.py]: Got it, the memoizer and the importer are now more loosely coupled.

File: corehq/apps/user_importer/importer.py (hunk 11 above)
Comment (snopoke) [reply corehq/apps/user_importer/importer.py]: Not in this PR. Possibly in another PR but only if it's an easy change.

**Review by snopoke — APPROVED** [review]

File: corehq/apps/users/util.py [hunk 15]
```diff
@@ -317,7 +317,7 @@ def _last_sync_needs_update(last_sync, sync_datetime):
 def log_user_change(by_domain, for_domain, couch_user, changed_by_user, changed_via=None,
                     change_messages=None, fields_changed=None, action=None,
                     by_domain_required_for_log=True, for_domain_required_for_log=True,
-                    bulk_upload_record_id=None):
+                    bulk_upload_record_id=None, group_changes_only=False):
```
Comment (mkangia) [inline corehq/apps/users/util.py:320]: Can you share why this change was needed?
I believe you should be able to log with just the change_messages and passing blank fields changed.

**Review by mkangia — COMMENTED** [review]
Thanks @Robert-Costello for the follow up and @snopoke for the close review. I have couple of blocking queries.

File: corehq/apps/users/util.py (hunk 15 above)
Comment (Robert-Costello) [reply corehq/apps/users/util.py:320]: You're right about that. Thanks, making the change.
